lib/Containers/MerkleTree.h

```diff
@@ -0,0 +1,259 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2019 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Dan Larkin-York
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGODB_CONTAINERS_MERKLE_TREE_H
+#define ARANGODB_CONTAINERS_MERKLE_TREE_H 1
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <memory>
+#include <mutex>
+#include <shared_mutex>
+#include <string>
+#include <string_view>
+#include <vector>
+
+namespace arangodb {
+namespace containers {
+
+template <std::size_t const BranchingBits = 3,  // 8 children per internal node,
+          std::size_t const LockStripes = 64  // number number of buckets that can be accessed simultaneously
+          >
+class MerkleTree {
+ protected:
+  static constexpr std::size_t CacheLineSize = 64;
```

Comment:
I would generally agree with this, but for two things. 1) We use an expected cache line size elsewhere in the code already and do not centralize it in lib. 2) I don't know that this is supported by any of our target compilers yet. It definitely isn't in g++-9.2.1.

Since it's not mission-critical that this value is 100% accurate, I'm just leaving a comment in the code to replace it in the future once it's available.